cmd/curator/curator_test.go

```diff
@@ -493,3 +586,301 @@ func TestHypershiftMonitorUpgrade(t *testing.T) {
+	}()
+
+	os.Args[1] = "intermediate-upgrade-cluster"
+	os.Args[2] = ClusterName
+
+	s.AddKnownTypes(clustercuratorv1.SchemeBuilder.GroupVersion, &clustercuratorv1.ClusterCurator{})
+	client := clientfake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(
+		getEUSUpgradeClusterCurator(),
+	).Build()
+
+	config, _ := config.LoadConfig("", "", "")
+
+	curatorRun(config, client, ClusterName, ClusterName)
+}
+
+func TestEUSFinalUpgrade(t *testing.T) {
+	// Test will fail because we don't have all the objects
+	// But that's ok, we just need to test the curator code
+	defer func() {
+		if r := recover(); r == nil {
+			t.Fatal("expected recover, but failed")
+		}
+	}()
+
+	os.Args[1] = "final-upgrade-cluster"
+	os.Args[2] = ClusterName
+
+	s.AddKnownTypes(clustercuratorv1.SchemeBuilder.GroupVersion, &clustercuratorv1.ClusterCurator{})
+	client := clientfake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(
+		getEUSUpgradeClusterCurator(),
+	).Build()
+
+	config, _ := config.LoadConfig("", "", "")
+
+	curatorRun(config, client, ClusterName, ClusterName)
+}
+
+func TestEUSMonitorUpgrade(t *testing.T) {
+	// Test will fail because we don't have all the objects
+	// But that's ok, we just need to test the curator code
+	defer func() {
+		if r := recover(); r == nil {
+			t.Fatal("expected recover, but failed")
+		}
+	}()
+
+	os.Args[1] = "intermediate-monitor-upgrade"
+	os.Args[2] = ClusterName
+
+	s.AddKnownTypes(clustercuratorv1.SchemeBuilder.GroupVersion, &clustercuratorv1.ClusterCurator{})
+	client := clientfake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(
+		getEUSUpgradeClusterCurator(),
+		getEUSClusterVersionManagedClusterView(),
+	).Build()
+
+	config, _ := config.LoadConfig("", "", "")
+
+	curatorRun(config, client, ClusterName, ClusterName)
+}
+
+func TestIntermediateUpdateImmutability(t *testing.T) {
+	clustercuratorv1.AddToScheme(scheme.Scheme)
+	m := &envtest.Environment{
+		CRDDirectoryPaths: []string{
+			filepath.Join("..", "..", "deploy", "crd"),
+		},
+	}
+
+	var err error
+	var cfg *rest.Config
+	if cfg, err = m.Start(); err != nil {
+		log.Fatal(err)
+	}
+
+	defer m.Stop()
+
+	var c client.Client
+
+	if c, err = client.New(cfg, client.Options{Scheme: scheme.Scheme}); err != nil {
+		log.Fatal(err)
+	}
+
+	err = c.Create(context.Background(), &corev1.Namespace{
+		ObjectMeta: v1.ObjectMeta{
+			Name: ClusterName,
+		},
+	})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = c.Create(context.Background(), getEUSUpgradeClusterCurator())
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	curator := clustercuratorv1.ClusterCurator{}
+	c.Get(context.TODO(), types.NamespacedName{
+		Namespace: ClusterName,
+		Name:      ClusterName,
+	}, &curator)
+
+	patch := []byte(
+		`{
+			"spec": {
+				"upgrade": {
+					"intermediateUpdate": "4.13.38"
+				}
+			}
+		}`)
+
+	err = c.Patch(context.Background(), &curator, client.RawPatch(types.MergePatchType, patch))
+
+	assert.NotNil(t, err, "intermediateUpdate immutable validation successful")
```

Comment:
It should've been an error with the message "The intermediateUpdate cannot be modified".
Can you get the curator again after this patch is run and see if the version is .38 or .37?

Reply:
@mikeshng Right, there is an error. Since I'm using assert.NotNil() it's expecting the error to not be nil. The test fails if there's no error.
![image](https://github.com/stolostron/cluster-curator-controller/assets/38960034/11882592-40f2-4760-abe7-18e8fbc48c07)

I added the check for version .37.

Reply:
I see. I think you should check the error string too so we know the error is the validation error we are expecting. Same comment for the other similar tests.

Reply:
@mikeshng Yep gotcha. I just put in those changes now https://github.com/stolostron/cluster-curator-controller/pull/212/commits/ddfdaaea6d6bec3bf5b0e0dc6dbf9a63e0e21d48